sympy/solvers/solvers.py

```diff
@@ -1320,29 +1334,41 @@ def _solve(f, *symbols, **flags):
                 soln = solve_undetermined_coeffs(f, symbols, ex, **flags)
             except NotImplementedError:
                 pass
-        if soln:
-            if flags.get('simplify', True):
-                if isinstance(soln, dict):
-                    for k in soln:
-                        soln[k] = simplify(soln[k])
-                elif isinstance(soln, list):
-                    if isinstance(soln[0], dict):
-                        for d in soln:
-                            for k in d:
-                                d[k] = simplify(d[k])
-                    elif isinstance(soln[0], tuple):
-                        soln = [tuple(simplify(i) for i in j) for j in soln]
+            if soln:
+                if flags.get('simplify', True):
+                    if isinstance(soln, dict):
+                        for k in soln:
+                            soln[k] = simplify(soln[k])
+                    elif isinstance(soln, list):
+                        if isinstance(soln[0], dict):
+                            for d in soln:
+                                for k in d:
+                                    d[k] = simplify(d[k])
+                        elif isinstance(soln[0], tuple):
+                            soln = [tuple(simplify(i) for i in j) for j in soln]
+                        else:
+                            raise TypeError('unrecognized args in list')
+                    elif isinstance(soln, tuple):
+                        sym, sols = soln
+                        soln = sym, {tuple(simplify(i) for i in j) for j in sols}
```

Comment:
It seems absurd to me that so many different cases need to be handled here. The code should really be refactored so that at least internally consistent types are used.

Reply:
I agree. The different return types for different types of equations is something that got baked in at the beginning and all development just retained those types. Anyone not liking this can use `dict=True`.  There's no getting around that *feature* until we agree that it's a good time to break backwards compatibilty. The docstring of `solve` enumerates all the different input and return types.

Reply:
I appreciate the significance of "internally" in your comment. It's just that at some point you have to return the expected type of the solution, regardless of what you do internally. Maybe it would be better to break out the simplification and checking process so the `_solve` routine doesn't have to display this level of complexity?

Reply:
The different types of input and output should ideally be handled in one function only before/after any other processing so that all the logic is in one place and internally a consistent form (presumably list of dicts) should be used everywhere.